src/kartograf/atom_mapper.py

```diff
@@ -855,6 +854,38 @@ def suggest_mapping_from_rdmols(
 
         return mapping
 
+    def _split_protein_components_molecules(self, protein: ProteinComponent) -> list[ProteinComponent]:
+        """
+        Aims at splitting a protein component into different protein components based on the
+        connectivity of the molecules that compose it. Useful for mapping multimer components
+        or proteins with structural waters or similarly.
+
+        This returns a ``ProteinComponent`` object with the name attribute indicating the starting
+        index in the original component.
+        """
+        from rdkit.Chem.rdmolops import GetMolFrags
+        rdmol = protein.to_rdkit()
+        fragments_indices = GetMolFrags(rdmol)
+        components = []
+        for index_tuple in fragments_indices:
+            edit_rdmol_frag = Chem.EditableMol(rdmol)
+            remove_indices = []
+            for atom in rdmol.GetAtoms():
+                atom_index = atom.GetIdx()
+                if not (atom_index in index_tuple):
+                    remove_indices.append(atom_index)
+            # Need to remove separately https://github.com/rdkit/rdkit/issues/1366
```

Comment:
Am I correct in understanding that this is because the atom ids get re-assigned on the fly?

From some pen and paper playing around, I think this is should work in all cases - are you reasonably confident of this too?

Reply:
Yes, that's my understanding. It is happening on the fly, so the iterator gets invalidated and the behavior is undefined.